File: src/main/java/org/tobi29/scapes/client/ScapesClient.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.Scapes;
import org.tobi29.scapes.client.input.InputMode;
import org.tobi29.scapes.client.input.gamepad.InputModeGamepad;
import org.tobi29.scapes.client.input.keyboard.InputModeKeyboard;
import org.tobi29.scapes.client.input.spi.InputModeProvider;
import org.tobi29.scapes.engine.Game;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.gui.GuiAlignment;
import org.tobi29.scapes.engine.gui.GuiComponentIcon;
import org.tobi29.scapes.engine.gui.GuiComponentText;
import org.tobi29.scapes.engine.gui.GuiNotification;
import org.tobi29.scapes.engine.input.Controller;
import org.tobi29.scapes.engine.input.ControllerDefault;
import org.tobi29.scapes.engine.input.ControllerJoystick;
import org.tobi29.scapes.engine.input.InputException;
import org.tobi29.scapes.engine.opengl.GL;
import org.tobi29.scapes.engine.utils.BufferCreator;
import org.tobi29.scapes.engine.utils.VersionUtil;
import org.tobi29.scapes.engine.utils.graphics.Image;
import org.tobi29.scapes.engine.utils.graphics.PNG;
import org.tobi29.scapes.engine.utils.io.filesystem.FileSystemContainer;
import org.tobi29.scapes.engine.utils.io.filesystem.classpath.ClasspathPath;
import org.tobi29.scapes.engine.utils.io.tag.TagStructure;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.Stream;

public class ScapesClient extends Game {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(ScapesClient.class);
    private final List<InputMode> inputModes = new ArrayList<>();
    private Image icon;
    private InputMode inputMode;
    private boolean freezeInputMode;

    private static Optional<InputMode> loadService(ScapesEngine engine,
            Controller controller, TagStructure tagStructure) {
        for (InputModeProvider provider : ServiceLoader
                .load(InputModeProvider.class)) {
            try {
                InputMode inputMode =
                        provider.get(engine, controller, tagStructure);
                if (inputMode != null) {
                    LOGGER.debug("Loaded input mode: {}",
                            provider.getClass().getName());
                    return Optional.of(inputMode);
                }
            } catch (ServiceConfigurationError e) {
                LOGGER.warn("Unable to load input mode provider: {}",
                        e.toString());
            }
        }
        if (controller instanceof ControllerDefault) {
            return Optional.of(new InputModeKeyboard(engine,
                    (ControllerDefault) controller, tagStructure));
        } else if (controller instanceof ControllerJoystick) {
            return Optional.of(new InputModeGamepad(engine,
                    (ControllerJoystick) controller, tagStructure));
        }
        return Optional.empty();
    }

    public InputMode inputMode() {
        return inputMode;
    }

    @Override
    public String name() {
        return "Scapes";
    }

    @Override
    public String id() {
        return "Scapes";
    }

    @Override
    public VersionUtil.Version version() {
        return Scapes.VERSION;
    }

    @Override
    public Image icon() {
        return icon;
    }

    @Override
    public void init() {
        try {
            Path path = engine.home();
            Path playlistsPath = path.resolve("playlists");
            Files.createDirectories(playlistsPath.resolve("day"));
            Files.createDirectories(playlistsPath.resolve("night"));
            Files.createDirectories(playlistsPath.resolve("battle"));
            Files.createDirectories(path.resolve("plugins"));
            Files.createDirectories(path.resolve("saves"));
            FileSystemContainer files = engine.files();
            files.registerFileSystem("Scapes",
                    new ClasspathPath(getClass().getClassLoader(),
                            "assets/scapes/tobi29/"));
            icon = files.get("Scapes:image/Icon.png").readReturn(
                    stream -> PNG.decode(stream, BufferCreator::bytes));
        } catch (IOException e) {
            engine.crash(e);
        }
        TagStructure tagStructure = engine.tagStructure();
        if (!tagStructure.has("Scapes")) {
            TagStructure scapesTag = tagStructure.getStructure("Scapes");
            scapesTag.setFloat("AnimationDistance", 0.15f);
            scapesTag.setBoolean("FXAA", true);
            scapesTag.setBoolean("Bloom", true);
            scapesTag.setDouble("RenderDistance", 128.0);
            TagStructure integratedServerTag =
                    scapesTag.getStructure("IntegratedServer");
            TagStructure serverTag = integratedServerTag.getStructure("Server");
            serverTag.setInteger("MaxLoadingRadius", 256);
            TagStructure socketTag = serverTag.getStructure("Socket");
            socketTag.setInteger("MaxPlayers", 5);
            socketTag.setInteger("WorkerCount", 1);
            socketTag.setInteger("RSASize", 1024);
        }
    }

    @Override
    public void initLate(GL gl) {
        loadInput();
    }

    @Override
    public void step() {
        if (engine.container().joysticksChanged()) {
            loadInput();
        }
        InputMode newInputMode = null;
        for (InputMode inputMode : inputModes) {
            if (inputMode.poll()) {
                newInputMode = inputMode;
            }
        }
        if (newInputMode != null && inputMode != newInputMode &&
                !freezeInputMode) {
            LOGGER.info("Setting input mode to {}", newInputMode);
            inputMode = newInputMode;
            engine.setGUIController(inputMode.guiController());
            GuiNotification message =
                    new GuiNotification(engine.globalGUI(), 500, 0, 290, 60,
                            GuiAlignment.RIGHT, 3.0);
            new GuiComponentIcon(message, 10, 10, 40, 40,
                    engine.graphics().textures()
                            .get("Scapes:image/gui/input/Default"));
            new GuiComponentText(message, 60, 25, 420, 10,
                    inputMode.toString());
        }
    }

    @Override
    public void dispose() {
    }

    public void loadInput() {
        LOGGER.info("Loading input");
        TagStructure tagStructure = engine.tagStructure().getStructure("Scapes")
                .getStructure("Input");
        Optional<InputMode> inputModeDefault =
                loadService(engine, engine.controller(), tagStructure);
        if (!inputModeDefault.isPresent()) {
            throw new InputException("No keyboard controller installed");
        }
        inputModes.clear();
        inputModeDefault.ifPresent(inputModes::add);
        for (ControllerJoystick joystick : engine.container().joysticks()) {
            loadService(engine, joystick, tagStructure)
                    .ifPresent(inputModes::add);
        }
        inputMode = inputModes.get(0);
        engine.setGUIController(inputMode.guiController());
    }

    public void setFreezeInputMode(boolean freezeInputMode) {
        this.freezeInputMode = freezeInputMode;
    }

    public Stream<InputMode> inputModes() {
        return inputModes.stream();
    }
}

File: ScapesFrontend/src/main/kotlin/org/tobi29/scapes/client/ScapesClient.kt
/*
 * Copyright 2012-2018 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.client

import kotlinx.coroutines.experimental.async
import org.tobi29.graphics.*
import org.tobi29.io.IOException
import org.tobi29.io.ReadableByteStream
import org.tobi29.io.filesystem.*
import org.tobi29.io.tag.*
import org.tobi29.logging.KLogging
import org.tobi29.scapes.engine.ComponentLifecycle
import org.tobi29.scapes.engine.ComponentStep
import org.tobi29.scapes.engine.ScapesEngine
import org.tobi29.scapes.plugins.Plugins
import org.tobi29.utils.ComponentTypeRegistered
import org.tobi29.utils.ComponentTypeRegisteredPermission
import kotlin.collections.set

class ScapesClient(
    val engine: ScapesEngine,
    val home: FilePath,
    savesSupplier: (ScapesClient) -> SaveStorage,
    lightDefaults: Boolean = false
) : ComponentLifecycle, ComponentStep {
    val configMap = engine[ScapesEngine.CONFIG_MAP_COMPONENT].mapMut("Scapes")
    val saves = savesSupplier(this)

    var animations by configMap.tagBoolean("Animations", !lightDefaults)
    var bloom by configMap.tagBoolean("Bloom", !lightDefaults)
    var autoExposure by configMap.tagBoolean("AutoExposure", !lightDefaults)
    var fxaa by configMap.tagBoolean("FXAA", !lightDefaults)
    var renderDistance by configMap.tagDouble(
        "RenderDistance",
        if (lightDefaults) 64.0 else 128.0
    )
    var resolutionMultiplier by configMap.tagDouble(
        "ResolutionMultiplier",
        if (lightDefaults) 0.5 else 1.0
    )

    override fun init(holder: ScapesEngine) {
        if (!configMap.containsKey("IntegratedServer")) {
            configMap["IntegratedServer"] = TagMap {
                this["Server"] = TagMap {
                    this["MaxLoadingRadius"] = 288.toTag()
                    this["Socket"] = TagMap {
                        this["MaxPlayers"] = 5.toTag()
                        this["WorkerCount"] = 1.toTag()
                    }
                }
            }
        }
        Plugins.setupAssets(engine.files)
        holder.registerComponent(Screenshots.COMPONENT, object : Screenshots {
            override fun saveScreenshot(image: Bitmap<*, *>) {
                val path = home.resolve("screenshots")
                    .resolve("${System.currentTimeMillis()}.png")
                try {
                    write(path) { encodePng(image, it, 9, false) }
                } catch (e: IOException) {
                    logger.warn(e) { "Error saving screenshot" }
                }
            }

            override fun screenshots(): List<Screenshots.Screenshot> = try {
                val path = home.resolve("screenshots")
                listRecursive(path) {
                    filter { isRegularFile(it) }
                        .filter { isNotHidden(it) }.toList()
                }.sorted()
            } catch (e: IOException) {
                logger.warn { "Failed to list screenshots: $e" }
                emptyList<FilePath>()
            }.map { file ->
                object : Screenshots.Screenshot {
                    override val image by lazy {
                        async(engine.taskExecutor) {
                            try {
                                decodePng(file)
                            } catch (e: IOException) {
                                logger.warn(e) { "Failed to decode screenshot" }
                                MutableIntByteViewBitmap(1, 1, RGBA)
                            }
                        }
                    }

                    override fun delete(): Boolean {
                        try {
                            delete(file)
                            return true
                        } catch (e: IOException) {
                            logger.warn(e) { "Failed to delete screenshot" }
                        }
                        return false
                    }
                }
            }
        })
    }

    override fun dispose(holder: ScapesEngine) {
        holder.unregisterComponent(Screenshots.COMPONENT)
    }

    companion object : KLogging() {
        val COMPONENT =
            ComponentTypeRegistered<ScapesEngine, ScapesClient, Any>()
    }
}

interface DialogProvider {
    fun openMusicDialog(result: (String, ReadableByteStream) -> Unit)

    fun openSkinDialog(result: (String, ReadableByteStream) -> Unit)

    fun saveScreenshotDialog(result: (FilePath) -> Unit)

    companion object {
        val COMPONENT =
            ComponentTypeRegisteredPermission<ScapesEngine, DialogProvider, Any>(
                "scapes.dialogs"
            )
    }
}
